src/jit/morph.cpp

```diff
@@ -1242,158 +1242,96 @@ void fgArgInfo::RemorphReset()
     nextSlotNum = INIT_ARG_STACK_SLOT;
 }
 
-fgArgTabEntry* fgArgInfo::RemorphRegArg(
-    unsigned argNum, GenTree* node, GenTree* parent, regNumber regNum, unsigned numRegs, unsigned alignment)
+fgArgTabEntry* fgArgInfo::UpdateRegArg(fgArgTabEntry* curArgTabEntry, GenTree* node, bool reMorphing)
 {
-    fgArgTabEntry* curArgTabEntry = nullptr;
-    unsigned       regArgInx      = 0;
-    unsigned       inx;
-
-    for (inx = 0; inx < argCount; inx++)
-    {
-        curArgTabEntry = argTable[inx];
-        if (curArgTabEntry->argNum == argNum)
-        {
-            break;
-        }
+    bool isLateArg = (curArgTabEntry->lateArgInx != (unsigned)-1);
+    assert((isLateArg && ((node->gtFlags & GTF_LATE_ARG) != 0)) ||
+           (!isLateArg && ((node->gtFlags & GTF_LATE_ARG) == 0)));
 
-        bool     isRegArg;
-        GenTree* argx;
-        if (curArgTabEntry->parent != nullptr)
-        {
-            assert(curArgTabEntry->parent->OperIsList());
-            argx     = curArgTabEntry->parent->Current();
-            isRegArg = (argx->gtFlags & GTF_LATE_ARG) != 0;
-        }
-        else
-        {
-            argx     = curArgTabEntry->node;
-            isRegArg = true;
-        }
-
-        if (isRegArg)
-        {
-            regArgInx++;
-        }
-    }
-    // if this was a nonstandard arg the table is definitive
-    if (curArgTabEntry->isNonStandard)
+    if (curArgTabEntry->parent != nullptr)
     {
-        regNum = curArgTabEntry->regNum;
+        assert(curArgTabEntry->parent->OperIsList());
+        assert(curArgTabEntry->parent->Current() == node);
     }
 
-    assert(curArgTabEntry->argNum == argNum);
-    assert(curArgTabEntry->regNum == regNum);
-    assert(curArgTabEntry->alignment == alignment);
-    assert(curArgTabEntry->parent == parent);
-
     if (curArgTabEntry->node != node)
     {
         GenTree* argx     = nullptr;
         unsigned regIndex = 0;
-
-        /* process the register argument list */
-        for (GenTreeArgList* list = callTree->gtCall.gtCallLateArgs; list; (regIndex++, list = list->Rest()))
+        if (reMorphing)
         {
-            argx = list->Current();
-            assert(!argx->IsArgPlaceHolderNode()); // No place holders nodes are in gtCallLateArgs;
-            if (regIndex == regArgInx)
+            // Find the arg in the late args list.
+            unsigned regArgInx = curArgTabEntry->lateArgInx;
+            for (GenTreeArgList* list = callTree->gtCall.gtCallLateArgs; list; (regIndex++, list = list->Rest()))
```

Comment:
Seems like this loop should be a member function: `GenTreeCall::GetLateArgByIdx()`

Reply:
Done, though I added it as a static member of `Compiler`, as that's where the other related methods are.